Segwit light

domob1812 opened the pull request:
This implements "segwit light" as described in the [design doc](https://docs.google.com/document/d/1EM2fO1kYksupEx-Mp-2Gi64IEWKM0KUyTVYqnnch3_0/edit?usp=sharing).  We create a new time-based fork (for now scheduled at block time 2'000'000'000, to be changed to a real timestamp later); all transactions made after the fork create UTXOs not indexed by their normal hash (txid), but instead by a hash not including any signature data (bare txid).  This means that all transactions made after the fork can be safely used in a chain of unconfirmed, pre-signed transactions and won't be affected by malleability.

domob1812 committed:
Regtest for segwit light.

This adds the new regtest segwit_light.py, which verifies that sending
of transactions and staking still works after the fork, and that the
new rules work as expected.

It also explicitly checks that transaction malleability is not an issue
anymore for chains of pre-signed transactions.

domob1812 committed:
Bare txid for GetTransaction / GetWalletTx.

This extends the two functions GetWalletTx and GetTransaction to (also)
find transactions by bare txid, not just by the normal txid.

These methods are mainly used in places where we need to look up e.g.
the previous transaction to a spend, so that we can know the address
that is being spent or the value of the input.

The change is fine to do (won't cause any extra consensus changes)
because all it does is make those methods return the correct previous
transaction (for after the fork) in cases where they would have
failed otherwise (since both are SHA-256d hashes and thus cannot have
collisions).  Also actual checks that some spent coin actually exists
are the explicitly on the consensus-level anyway.

domob1812 committed:
Use outputhash in regtests.

This updates the regtests to use "outputhash" for listunspent results
in some places, to make sure the code will also work after activating
segwit-light.

Some other places remain where outputs are not from listunspent and that
still needs to be updated when segwit-light gets activated generally,
but this is a first step to reduce the amount of required changes then.

domob1812 commented:
This needs #59 and #60 merged first.

domob1812 commented:
Even after activating the fork, almost all regtests still pass.  Only the `around_segwit_light.py` test fails, but that is expected since it tries to explicitly activate the fork and test pre- and post-fork behaviour.  Some of the other tests need trivial changes as in 5b29b79e55f251b070e179b3e9532202822509c8.

Before merging, that last commit should be removed from the PR and just applied back after fork activation (so the tests then still pass).

domob1812 commented:
This is now ready for review.

galpHub commented on divi/src/BlockMemoryPoolTransactionCollector.cpp:
```diff
@@ -18,10 +20,6 @@ extern CCoinsViewCache* pcoinsTip;
 extern bool fDebug;
 
 class CScriptCheck;
-unsigned int GetLegacySigOpCount(const CTransaction& tx);
-unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& mapInputs);
-bool CheckInputs(const CTransaction& tx, CValidationState& state, const CCoinsViewCache& view, bool fScriptChecks, unsigned int flags, bool cacheStore, std::vector<CScriptCheck>* pvChecks = NULL );
-void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo& txundo, int nHeight);
```
Instead of including main to get access to these functions, please separate them out into a separate file to avoid recompiling main each time.

domob1812 replied:
I personally disagree with this, as I'd rather have longer compilation time than duplicated forward declarations (which may introduce bugs, although the risk is admittedly rather low).  But as this is your decision, I've reverted that part of the change now and updated the forward declarations instead of including `main.h`.

It may make sense to move some of these functions to a separate file and header as you say (e.g. upstream Bitcoin has long created a `validation` file), but that would be a separate PR.  I can do that before this if you want, or we just keep it in mind for after.

galpHub replied:
I agree with your concern over duplicated forward declarations of functions. I would also rather include the header directly __if there were no alternative allowing moving the definition out of main__. Moving the necessary data into header/source pairs ought to be preferred. There's no need to pay for a long compilation time where a shorter one with the same amount of bugs would do. If you have concerns over potential name resolution I'm all for placing them either in their own class or namespace to avoid those potential sources of conflict.

domob1812 replied:
Makes sense.  I saw you did some code moves recently.  I will rebase this PR on top of them, and then see what to do here.  In any case, I think those moves can/should be separate PRs merged before this one then.

domob1812 replied:
Rebased & thanks to your work, this is now obsolete :+1:

galpHub commented on divi/src/ForkActivation.cpp:
```diff
@@ -25,12 +25,15 @@ const std::unordered_map<Fork, int64_t,std::hash<int>> ACTIVATION_TIMES = {
 } // anonymous namespace
 
 ActivationState::ActivationState(const CBlockIndex* pi)
-  : nTime(pi->nTime)
+  : nTime(pi == nullptr ? 0 : pi->nTime)
 {}
 
```
Should default to no-activation, to avoid accidental misuse

domob1812 replied:
That's what it does.  Time zero means no fork will be activated.

galpHub commented on divi/src/main.cpp:
```diff
@@ -1357,20 +1358,23 @@ void static InvalidBlockFound(CBlockIndex* pindex, const CValidationState& state
     }
 }
 
-void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo& txundo, int nHeight)
+void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo& txundo,
+                 const ActivationState& as, const int nHeight)
 {
     // mark inputs spent
     if (!tx.IsCoinBase() ) {
         txundo.vprevout.reserve(tx.vin.size());
-        BOOST_FOREACH (const CTxIn& txin, tx.vin) {
+        for (const auto& txin : tx.vin) {
             txundo.vprevout.push_back(CTxInUndo());
             bool ret = inputs.ModifyCoins(txin.prevout.hash)->Spend(txin.prevout.n, txundo.vprevout.back());
             assert(ret);
         }
     }
 
     // add outputs
-    inputs.ModifyCoins(tx.GetHash())->FromTx(tx, nHeight);
+    const uint256 outputHash = (as.IsActive(Fork::SegwitLight)
+                                  ? tx.GetBareTxid() : tx.GetHash());
+    inputs.ModifyCoins(outputHash)->FromTx(tx, nHeight);
```
This doesn't seem very safe from an upgradability perspective. It seems like it is possible that an older transaction that was formerly referred to by legacy hash will now fail to be found by updated hash

domob1812 replied:
I'm not sure I understand your concern here.  A transaction will always be referred to based on the activation state at the block it was confirmed in.  There's no ambiguity there, and also e.g. the wallet will be able to properly handle spending any coins, created before or after the fork.

Perhaps a "weird" thing that could happen is if there is a reorg right around the fork, then a transaction might have been confirmed with e.g. old rules on one fork, and with new rules on another.  But even if that happens (which seems really unlikely in practice), that wouldn't cause any issues anywhere.

galpHub commented on divi/src/main.cpp:
```diff
@@ -1540,8 +1546,11 @@ bool DisconnectBlock(CBlock& block, CValidationState& state, CBlockIndex* pindex
         // specially with outsEmpty.
         {
+            const uint256 outputHash = (as.IsActive(Fork::SegwitLight)
+                                          ? tx.GetBareTxid() : tx.GetHash());
+
             CCoins outsEmpty;
-            CCoinsModifier outs = view.ModifyCoins(hash);
+            CCoinsModifier outs = view.ModifyCoins(outputHash);
```
Same concern as above

domob1812 replied:
This is indeed part of "the same" change as above, so I'll close this discussion thread here (as it will always have to be in sync with the other place anyway).

galpHub commented on divi/src/txmempool.cpp:
```diff
@@ -577,7 +580,7 @@ void CTxMemPool::remove(const CTransaction& origTx, std::list<CTransaction>& rem
             // happen during chain re-orgs if origTx isn't re-accepted into
             // the mempool for any reason.
             for (unsigned int i = 0; i < origTx.vout.size(); i++) {
-                std::map<COutPoint, CInPoint>::iterator it = mapNextTx.find(COutPoint(origTx.GetHash(), i));
+                auto it = mapNextTx.find(COutPoint(segwitLight ? origTx.GetBareTxid() : origTx.GetHash(), i));
                 if (it == mapNextTx.end())
```
Same as above

domob1812 replied:
Mempool logic is a bit broken around the fork, indeed.  But with the restriction we have put in place, it will be fine in practice.

For this particular place here (and the next one below), we could of course just try to look up both hashes and use the one that works (if any).  Is that something you would prefer?

galpHub commented on divi/src/txmempool.cpp:
```diff
@@ -591,7 +594,7 @@ void CTxMemPool::remove(const CTransaction& origTx, std::list<CTransaction>& rem
             const CTransaction& tx = mapTx[hash].GetTx();
             if (fRecursive) {
                 for (unsigned int i = 0; i < tx.vout.size(); i++) {
-                    std::map<COutPoint, CInPoint>::iterator it = mapNextTx.find(COutPoint(hash, i));
+                    auto it = mapNextTx.find(COutPoint(segwitLight ? tx.GetBareTxid() : tx.GetHash(), i));
```
Same as above

galpHub commented on divi/src/txmempool.cpp:
```diff
@@ -703,7 +713,8 @@ void CTxMemPool::check(const CCoinsViewCache* pcoins) const
             CTransaction tx2;
-            if (lookup(txin.prevout.hash, tx2)) {
+            if ((segwitLight && lookupBareTxid(txin.prevout.hash, tx2))
+                  || (!segwitLight && lookup(txin.prevout.hash, tx2))) {
```
In the event of timestamps not being monotonic, a transaction that is saved in one block under 'segwit-light' will be 'not found' if the next block is not 'segwit-light'

domob1812 replied:
I can't find this place in the code anymore, but I guess what you mean here is that the mempool logic can be inconsistent around fork activation.  This is the case independent of timestamps being monotonic or not, and is why we have this special rule of forbidding spending of unconfirmed outputs in the mempool policy (and wallet) for a time window around the fork.  This time window should be large enough to make sure there are no issues like this in practice.

On the core consensus level, non-monotonic timestamps are of course fine as well.  For a transaction confirmed in a particular block, that block's timestamp (independent of what came before and comes after) determines what UTXOs get created, and the wallet later also handles this correctly when spending those UTXOs again.

galpHub commented on divi/src/Makefile.am:
```diff
@@ -584,7 +584,7 @@ divid_LDADD += $(LIBBITCOIN_ZMQ) $(ZMQ_LIBS)
 endif
 
 if ENABLE_WALLET
-divid_LDADD += libbitcoin_wallet.a \
+divid_LDADD += $(LIBBITCOIN_WALLET) \
   $(LIBBITCOIN_CRYPTO)
```
We  could pull these makefile changes into their own separate PR since it's clear they're separate from the PR.

domob1812 replied:
Good point, I've split this out (and extended it) in #64.  The reason why this was part of this PR is that it needed more extensive changes at a previous state of the PR to make it build, and this part (which is indeed not specific to the current PR) remained.

galpHub commented on divi/src/wallet.h:
```diff
@@ -178,6 +178,11 @@ class CWallet : public CCryptoKeyStore, public NotificationInterface, public I_K
 
+    /** Returns true if the wallet should allow spending of unconfirmed change.
+     *  This is mostly determined by allowSpendingZeroConfirmationOutputs,
+     *  but is forced to off around the segwit-light fork.  */
+    bool AllowSpendingZeroConfirmationChange() const;
```
Wouldn't it be simpler to do the upgrade process in two separate upgrades?
(1) You change the GetHash function to return the legacy hash if the transaction version is some specific value (the present default) and to return the bare txid for the upgraded version number.
(2) You upgrade the network as a soft fork by not including the updated version number directly.
(3) Once everybody has made the first update, the second update is a softfork relative to the first one.

What are the limitations to doing this?

domob1812 replied:
That is an interesting suggestion.  I don't see any reason why it wouldn't work.  However, I believe it would basically just make things more complex without much benefit:
- it would still be a hard fork (one of them, namely step 1), so no benefit in this respect
- it would be two forks instead of one that need to be scheduled, people notified, and hoping that they update in time, which I think rather increases than decreases the overall risk
- instead of a rather simple rule change "in one place" (concerning consensus at least), namely what UTXOs get created for a particular transaction, we now have a more complex dependency between the transaction version, things the version implies, and consensus rules about what versions are allowed

I can see one benefit of this change, and that is fixing the spending of unconfirmed outputs in a time-window "around" fork activation (which is probably why you brought up this suggestion here).  However, I would much rather have some temporary logic to avoid issues as in this PR (which, notably, also does not affect the core consensus at all) than more complicated fork logic in general that we need to keep for the future.

galpHub replied:
Some clarification on item (1), you wouldn't update the version number in step (1) so that it remains a soft-fork. You'd update it at a later stage after you've ensured enough people have upgraded. Blocks with the hard fork active would only accept newer-version transactions perhaps? Sounds messier I agree. I'm not entirely certain about the best way to go about it.

domob1812 replied:
I've thought about this a bit more.  In fact, I think it needs three forks, which at least conceptually are separate (but of course they could under some circumstances be part of the same code update, just scheduled at different times):

1. A soft fork that disallows a certain transaction version completely (or all versions larger than current default).
2. A hard fork that changes what UTXOs get generated for transactions with a new version (which aren't valid on the network at this stage at all, so this change is rather easy to do after the network has updated and activated step 1 successfully).
3. Another hard fork that actually allows transactions of the new version again.

Step 1 is relatively uncritical as a straight-forward soft fork.  But once step 1 is activated on the network, steps 2 & 3 could be done in a follow-up with perhaps a bit less risk and a combined update.  They remain a hard fork, though.

I think this process could work, and might be able to reduce the upgrade risk.  But I'm still not fully convinced it is the best way forward; the only real benefit would be removing that temporary restriction of spending unconfirmed outputs around the activation of the current PR.  And the version-number-dependent logic in itself is then something a bit more complex that remains forever.

domob1812 committed:
Restrict unconfirmed change around fork.

Around the segwit-light fork (some hours before and after), we should
avoid spending unconfirmed outputs:  If we do that, it is unclear
whether or not the fork will be active in the block that confirms
those spends, and thus we can't reliably construct a valid chain.

With this change, we introduce a (temporary) measure to disallow
those spends in the wallet and mempool policy in a window of time around
the planned fork activation.

domob1812 committed:
Add TransactionUtxoHasher abstraction.

Some places in the code need to determine the hash by which the UTXOs
of a given transaction will be referred to; in particular, we need that
when processing UTXO set updates for block connects and disconnects, in
the mempool and for assembling transactions into a new block.

This commit introduces a class TransactionUtxoHasher, which abstracts
this step and is used in all those places in the code instead of just
getting the txid directly.

For now, this has no effects on behaviour; but it makes it more clear in
the code where we need this particular logical feature; it will allow us
to add some more unit tests for those parts with explicit mocks of the
hasher class; and it will make it easier to implement segwit-light in the
future (where we basically just need to flip the hasher implementation but
no other parts in the code).

domob1812 commented:
I've started to work on splitting out some more parts of this PR.  Marking as draft in the mean time (but the discussions started here can of course continue).

domob1812 committed:
Unit test for UpdateCoins UTXO creation.

Add a unit test (together with the necessary framework around) for
UTXO creation from UpdateCoins, based on the UTXO hasher (rather than
the txid directly).

domob1812 committed:
Unit test for mempool with UTXO hasher.

This extends the mempool unit tests to explicitly verify that
adding transactions, removing transactions, checking the pool
and looking up coins / transactions still works even if we use
the bare txid for some transactions as UTXO hash (as will be
the case with segwit-light in the future).

domob1812 committed:
Consensus and mempool parts of segwit-light.

This implements "segwit-light" in the node (consensus and mempool):
After activation of a new fork, outputs created by new transactions
are referred to by the creating transaction's bare txid rather than
the normal hash (txid).  This makes chains of constructed transactions
immune to transaction malleability.

In the mempool, we use the current chain tip to determine the activation
state of the fork, as we can't know for sure when a transaction will be
confirmed (and thus what activation state will be in effect then).  This
will lead to issues right around the fork activation time, but we will
simply disallow spending of unconfirmed outputs in the mempool and
wallet "around" the fork time temporarily to resolve this.

The wallet is also not yet updated to take the change into account
when selecting coins.

domob1812 committed:
Use UTXO hasher for spending coins.

Use the UTXO hasher abstraction in the wallet and for staking (i.e.
for places where coins are spent).  The wallet gets its own instance,
which will allow for dependency injection in tests.

For now, the hasher used in the wallet is just the normal hasher, i.e.
there are no actual changes in behaviour.  In the future, the wallet
hasher can be changed accordingly for the activation of segwit light.

domob1812 committed:
Segwit light for spending coins.

Handle segwit-light properly when spending coins, either for staking
or normal transactions in the wallet.  Depending on when a transaction
was confirmed, we need to make sure to include the right outpoint (with
txid or bare txid) in the transaction spending an output.  This is done
through a custom UTXO hasher used in the wallet, which specifically
works for CWalletTx.

domob1812 committed:
Test UTXO hasher in PoSTransactionCreator.

This adds basic unit tests for PoSTransactionCreator, with a fake
environment that allows a real PoSTransactionCreator to create
staking transactions.

In particular, we use that to unit test that it uses the UTXO hasher
of the wallet properly.

domob1812 committed:
Unit tests for UTXO hasher in wallet.

Extend the unit tests in wallet_coinmanagement_tests.cpp to include
also explicit checks for situations in which the wallet is supposed
to the UTXO hasher rather than e.g. a plain transaction hash.

domob1812 committed:
Move internal functions into anonymous namespace.

Wrap some of the functions used only in main into an anonymous namespace;
some of them were already marked as "static", others were not (but they
were still not needed from anywhere else).

This makes it immediately clear that those functions are only used in the
context of main.

FindUndoPos had a declaration first and the definition later, but it is
fine to just move the definition to where the declaration was.  This
simplifies the code further (and is a pure move in the context of
this commit).

domob1812 committed:
Add bare txid to transaction index.

When the transaction index is enabled (-txindex), this change also keeps
an index from bare txid to the disk position, so that lookups with the
index work on both normal txids and bare txids.

With this, the previous change made to GetTransaction will be improved
for enabled -txindex, and it will also allow us to look up any transaction
by bare txid, even if the fallbacks in GetTransaction fail.

domob1812 committed:
Update txindex.py test for bare txid lookups.

This extends the txindex.py test, to check lookups also by bare txid
where this is possible (mempool and txindex).  Lookups by UTXO set only
work on whatever hash is used for the UTXO before/after segwit light.

domob1812 committed:
Clean up some mn config code.

This is a bunch of related and straight-forward cleanups to the code
around masternode configuration:  Pass arguments as const& instead of
by value, mark functions as const, and move helper functions from
being static in a class and declared inside the header to just being
inside an anonymous namespace in the cpp file.

domob1812 commented:
Reopened in #82 with corrected updated branches.
